Use counter for tracking container states

cpuguy83 opened the pull request:
Container state counts are used for reporting in the `/info` endpoint.
Currently when `/info` is called, each container is iterated over and
the containers 'StateString()' is called. This is not very efficient
with lots of containers, and is also racey since `StateString()` is not
using a mutex and the mutex is not otherwise locked.

We could just lock the container mutex, but this is proven to be
problematic since there are frequent deadlock scenarios and we should
always have the `/info` endpoint available since this endpoint is used
to get general information about the docker host.

Closes #28912

dnephin commented on container/state.go:
```diff
@@ -301,6 +323,21 @@ func (s *State) IsPaused() bool {
+// SetPaused sets the container state to paused without locking.
+func (s *State) SetPaused() {
+	oldState := s.StateString()
+	defer func() { s.counter.SetCount(oldState, s.StateString()) }()
```
I think you could do something like:

```go
defer s.UpdateCounter()()
```

```go

func (s *State) UpdateCounter() func() {
    oldState := s.StateString()
    return func() {
        s.counter.SetCount(oldState, s.StateString())
    }
}

dnephin commented on container/state.go:
```diff
@@ -324,13 +364,17 @@ func (s *State) SetRemovalInProgress() bool {
+	oldState := s.StateString()
+	defer func() { s.counter.SetCount(oldState, s.StateString()) }()
 	s.Dead = true
 	s.Unlock()
```
Isn't this still racy if the container changes state between `oldState := s.StateString()` and the defer running?

It would reduce the count of the old state twice?

cpuguy83 replied:
Ah, you're right the `SetCount` needs to be done while the lock is held.

dnephin reviewed (commented):
Generally design SGTM, but I do have a question about it

stevvooe commented:
@cpuguy83 Is there any way we could leverage the prometheus metrics registry to aggregate this data?

AkihiroSuda commented:
The lock-free proposal (https://github.com/docker/docker/issues/30225) can close this PR?

cpuguy83 commented:
@AkihiroSuda I don't think so. Even with a lock-free container, it's still horrible to iterate every single container for `docker info`

stevvooe commented:
@AkihiroSuda I agree with @cpuguy83 here. Aggregating the entire dataset will be much less reliable than pre-aggregating with counters.

@cpuguy83 Any luck getting `go-metrics` to do this? Let me know if you need pointers there.

LK4D4 commented:
@cpuguy83 CI fails :/

cpuguy83 committed:
Use counter for tracking container states

Container state counts are used for reporting in the `/info` endpoint.
Currently when `/info` is called, each container is iterated over and
the containers 'StateString()' is called. This is not very efficient
with lots of containers, and is also racey since `StateString()` is not
using a mutex and the mutex is not otherwise locked.

We could just lock the container mutex, but this is proven to be
problematic since there are frequent deadlock scenarios and we should
always have the `/info` endpoint available since this endpoint is used
to get general information about the docker host.

Really, these metrics on `/info` should be deprecated. But until then,
we can just keep a running tally in memory for each of the reported
states.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
@stevvooe I have wired in go-metrics here, although I did have to keep a separate counter for reporting to `info` as there doesn't really seem to be a way to query these metrics in a desirable way.

This still has some issue somewhere that so far I've only seen pop in tests where the counts get way off that I need to figure out.

stevvooe commented:
> as there doesn't really seem to be a way to query these metrics in a desirable way.

@cpuguy83 That is a wise decision!

@crosbymichael @juliusv Any ideas here?

juliusv commented:
> Any ideas here?

There's indeed no way to get the value of a single metric with Prometheus's Go client library, except for calling `Write` on it to dump it to its protobuf representation.

If you're going to maintain the count yourself anyways, you may want to think about not tracking the state twice, but using throw-away `ConstMetrics` every time you are scraped that just proxy your internal state to Prometheus. This is the common pattern for exporters which just translate existing values to Prometheus metrics. See for example https://github.com/prometheus/memcached_exporter/blob/master/main.go#L334.

vdemeester commented:
ping @cpuguy83, what is the status here ? 👼

cpuguy83 commented:
Working on it.

cpuguy83 commented:
Running the tests a few times, the counts seem to be wrong by a predictable amount at this point.
I've tried a few different things to resolve this but can't seem to pinpoint the root issue... could use a fresh set of eyes.

@tonistiigi Do you see anything inherently wrong with the state tracking?

dnephin commented:
I've been running these tests locally and I can't get them to fail

cpuguy83 commented:
The whole suite needs to run.

AkihiroSuda commented on daemon/metrics.go:
```diff
@@ -40,3 +49,62 @@ func init() {
+		return
+	}
+	switch label {
+	case "running":
```
Why not use const enum?

cpuguy83 replied:
Mostly because the intention was to use this for prom which needs a meaningful label value where an enum doesn't cut it.

AkihiroSuda commented on daemon/monitor.go:
```diff
@@ -54,6 +54,9 @@ func (daemon *Daemon) StateChanged(id string, e libcontainerd.StateInfo) error {
 			defer autoRemove()
 		}
 
+		stateCtr.dec("running")
```
Don't we need to wrap such `dec-then-inc` with mutex?
If `docker info` is called during dec and inc, it would result in weird output.

cpuguy83 replied:
Technically yes, but at the same time I'm not too worried about it.
It's already much more racey than this.

rogaha commented:
@cpuguy83 it looks like the `containerd` maintainers are also working adding these types of metrics. @crosbymichael do you mind sharing with us the `approach` you guys are using?

cpuguy83 commented:
This is all green folks.

cpuguy83 commented:
Something to consider b/c once metrics are out of experimental we probably ought not change this...

`docker info` looks at running, paused, and stopped. Where if it's not running or paused it's considered stopped (vs created, exited, etc).
Should we break these values down better for metrics?

dnephin reviewed (approved):
LGTM

What fixed the counter problems? Was it changing the counters to be per-container containers?

cpuguy83 commented:
> What fixed the counter problems?

Just started stashing the actual container state when it gets updated rather than calling inc+dec on split states.
Then we build the summary when `info` or the metrics endpoint is called.

stevvooe commented:
LGTM after providing example `/metrics` output (in a gist or something).

@crosbymichael PTAL

cpuguy83 commented:
Example of the  metrics:

https://gist.github.com/cpuguy83/a04005ad22f03a46c5a0c95fe1d53d26

stevvooe commented:
@cpuguy83 Wondering if we should make an effort to drop the `engine_daemon_` prefixes... these seem redundant.

crosbymichael commented:
LGTM

Its the best we can do without the ability to query prom metrics in process. , @cpuguy83 not saying your implementation is bad, just working with what we have :)  It would be a lot less code if we could query the metric in process and get the values.

cpuguy83 commented:
> It would be a lot less code if we could query the metric in process and get the values

Can you explain what you mean by this?

crosbymichael commented:
If we didn't have to query for the info endpoint, you just have a single metric with labels, and you inc/dec the id+state for each of the states, its like 10 lines of code.

cpuguy83 commented:
@stevvooe Yeah, we could probably simplify the metrics namespace.
Maybe just `docker_engine_*` or just `docker_*`

stevvooe commented:
@cpuguy83 Right now, it is `engine_daemon_`, which is a bit redundant. I am unsure about putting `docker_` there, as well, since this is moby.

cpuguy83 commented:
@stevvooe Do you have a suggestion? Just remove all-together?

stevvooe commented:
> @stevvooe Do you have a suggestion? Just remove all-together?

I think we can remove the prefix, but I'd like to hear @crosbymichael's opinion. We put work into "namespacing" metrics but prometheus already has a lot of context, like "job", "target" and info structs that can be used to create that context.

rogaha commented:
LGTM

cpuguy83 commented:
@stevvooe Are you ok if we merge this and talk about namespace separately?

stevvooe commented:
LGTM

Let's open a bug to review the namespacing for metrics.

cpuguy83 commented:
Opened #33056 to discuss.

Merging this one since it has all the required LGTM's and is all green.
